File: app/src/main/java/com/ewaldmire/osmride/weight/WaistRepository.kt
package com.ewaldmire.osmride.weight

import android.content.Context
import java.io.File
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Persists logged waist measurements to app-private storage - same JSON-index pattern as
 * WeightRepository, just its own subdirectory. */
class WaistRepository(context: Context) {
    private val appContext = context.applicationContext
    private val waistDir: File = File(appContext.filesDir, "waist").apply { mkdirs() }
    private val indexFile = File(waistDir, "index.json")
    private val json = Json { ignoreUnknownKeys = true }

    private val _entries = MutableStateFlow(loadIndex())
    /** Newest first. */
    val entries: StateFlow<List<WaistEntry>> = _entries.asStateFlow()

    suspend fun addEntry(waistCm: Double, recordedAtEpochMillis: Long = System.currentTimeMillis()): WaistEntry =
        withContext(Dispatchers.IO) {
            val entry = WaistEntry(
                id = UUID.randomUUID().toString(),
                waistCm = waistCm,
                recordedAtEpochMillis = recordedAtEpochMillis,
            )
            val updated = (_entries.value + entry).sortedByDescending { it.recordedAtEpochMillis }
            _entries.value = updated
            saveIndex(updated)
            entry
        }

    suspend fun deleteEntry(id: String) = withContext(Dispatchers.IO) {
        val updated = _entries.value.filterNot { it.id == id }
        _entries.value = updated
        saveIndex(updated)
    }

    private fun loadIndex(): List<WaistEntry> {
        if (!indexFile.exists()) return emptyList()
        return try {
            json.decodeFromString<List<WaistEntry>>(indexFile.readText())
        } catch (e: Exception) {
            emptyList()
        }
    }

    private fun saveIndex(entries: List<WaistEntry>) {
        indexFile.writeText(json.encodeToString(entries))
    }
}

File: linux-client/osm_ride_linux/weight/waist_repository.py
"""Persists logged waist entries to local storage, same JSON-index pattern as
weight/repository.py (deliberately duplicated rather than shared - see WaistRepository.kt).

Mirrors app/src/main/java/com/ewaldmire/osmride/weight/WaistRepository.kt.
"""

from __future__ import annotations

import json
import os
import time
import uuid
from collections.abc import Callable
from dataclasses import asdict
from pathlib import Path

from .models import WaistEntry


def _data_home() -> Path:
    xdg = os.environ.get("XDG_DATA_HOME")
    base = Path(xdg) if xdg else Path.home() / ".local" / "share"
    return base / "osm-ride-linux"


class WaistRepository:
    def __init__(self, data_dir: Path | None = None) -> None:
        self._waist_dir = (data_dir or _data_home()) / "waist"
        self._waist_dir.mkdir(parents=True, exist_ok=True)
        self._index_file = self._waist_dir / "index.json"

        self.on_entries_changed: Callable[[list[WaistEntry]], None] | None = None
        # Newest first.
        self.entries: list[WaistEntry] = self._load_index()

    def add_entry(self, waist_cm: float, recorded_at_epoch_millis: int | None = None) -> WaistEntry:
        entry = WaistEntry(
            id=str(uuid.uuid4()),
            waist_cm=waist_cm,
            recorded_at_epoch_millis=recorded_at_epoch_millis or int(time.time() * 1000),
        )
        updated = sorted([*self.entries, entry], key=lambda e: e.recorded_at_epoch_millis, reverse=True)
        self._update_entries(updated)
        return entry

    def delete_entry(self, entry_id: str) -> None:
        updated = [e for e in self.entries if e.id != entry_id]
        self._update_entries(updated)

    def _update_entries(self, updated: list[WaistEntry]) -> None:
        self.entries = updated
        self._save_index(updated)
        if self.on_entries_changed:
            self.on_entries_changed(updated)

    def _load_index(self) -> list[WaistEntry]:
        if not self._index_file.exists():
            return []
        try:
            raw = json.loads(self._index_file.read_text(encoding="utf-8"))
            return [WaistEntry(**entry) for entry in raw]
        except (json.JSONDecodeError, KeyError, TypeError):
            return []

    def _save_index(self, entries: list[WaistEntry]) -> None:
        self._index_file.write_text(json.dumps([asdict(e) for e in entries], indent=2), encoding="utf-8")
